crates/pep508-rs/src/marker/algebra.rs

```diff
@@ -748,6 +751,110 @@ impl Edges {
     }
 }
 
+// Normalize a [`VersionSpecifier`] before adding it to the tree.
+fn normalize_specifier(specifier: VersionSpecifier) -> VersionSpecifier {
+    let (operator, version) = specifier.into_parts();
+
+    // The decision diagram relies on the assumption that the negation of a marker tree is
+    // the complement of the marker space. However, pre-release versions violate this assumption.
+    // For example, the marker `python_full_version > '3.9' or python_full_version <= '3.9'`
+    // does not match `python_full_version == 3.9.0a0`. However, it's negation,
+    // `python_full_version > '3.9' and python_full_version <= '3.9'` also does not include
+    // `3.9.0a0`, and is actually `false`.
+    //
+    // For this reason we ignore pre-release versions entirely when evaluating markers.
+    // Note that `python_version` cannot take on pre-release values so this is necessary for
+    // simplifying ranges, but for `python_full_version` this decision is a semantic change.
+    let mut release = version.release();
+
+    // Strip any trailing `0`s.
+    //
+    // The [`Version`] type ignores trailing `0`s for equality, but still preserves them in it's
+    // [`Display`] output. We must normalize all versions by stripping trailing `0`s to remove the
+    // distinction between versions like `3.9` and `3.9.0`, whose output will depend on which form
+    // was added to the global marker interner first.
```

Comment:
Maybe you mean
```suggestion
    // [`Display`] output. We must normalize all versions by stripping trailing `0`s to remove the
    // distinction between versions like `3.9` and `3.9.0`, otherwise the output would depend on which form
    // was added to the global marker interner first.
```